changelog.d/7289.bugfix

```diff
@@ -0,0 +1 @@
+Fix an edge-case where it was not possible to cross-sign a user which did not share a room with any user on your homeserver. The bug only affected Synapse deployments in worker mode.
```

Comment:
For want of a better place to write this: this seems to be in conflict with what Dave is saying:

>  I've been creating a direct chat with the user, accepting, then starting a verification

surely in that case the users *do* share a room?

I'm mostly concerned that we don't really understand what the bug is, and hence may not be fixing the right bug.

Relatedly: if this only affects worker-mode: doesn't that suggest that the real bug is to do with synchronisation between workers, rather than something we ought to be fixing at the federation layer?

Reply:
You're right in both respects. Dave did mention that a room had been created and the client did wait for this to complete before uploading key signatures.

However the problem lies in that [`incoming_signing_key_update`](https://github.com/matrix-org/synapse/blob/fef82f4e22d06bf914b5f5eb78ee7a09f610e63a/synapse/handlers/e2e_keys.py#L1126) still bails out at the `get_rooms_for_user`, according to Dave, when it should be discovering the DM room that got created.

So the real fix is to figure out why that function is returning 0. My *guess* is that the cache on this worker isn't being invalidated correctly when the DM room is created.

However, in both the interest of time and in terms of helping people who's homeservers already rejected the signing key update, this PR should help. But I definitely don't propose it as a long-term solution.

Reply:
The bug (supposedly) originates from this code:

https://github.com/matrix-org/synapse/blob/72fe2affb6ac86d433b80b6452da57052365aa26/synapse/handlers/e2e_keys.py#L1263-L1267

I guess this would also be triggered for a local user if they haven't joined any rooms at all.

Reply:
Well, in a monolith the DM that's created would be detected correctly, so then "did not share a room with any user on your homeserver" wouldn't make sense.

How about just:

Fix a bug with cross-signing devices of users on other homeservers while in worker mode.

Reply:
:man_shrugging: ok fine!